File: 022_Zigzag Conversion.cpp
/*
Zigzag Conversion

Convert s into zigzag form on numRows rows and then read line by line.

Strategy: Simulate rows and append characters moving down/up between rows.
*/

#include <string>
#include <vector>
#include <iostream>
using namespace std;

string convert(string s, int numRows) {
    if (numRows <= 1 || numRows >= (int)s.size()) return s;
    vector<string> rows(min(numRows, int(s.size())));
    int cur = 0;
    bool down = true;
    for (char c : s) {
        rows[cur] += c;
        if (cur == 0) down = true;
        else if (cur == numRows - 1) down = false;
        cur += down ? 1 : -1;
    }
    string res;
    for (auto &r : rows) res += r;
    return res;
}

int main() {
    cout << "Zigzag 1: " << convert("PAYPALISHIRING", 3) << '\n'; // PAHNAPLSIIGYIR
    cout << "Zigzag 2: " << convert("PAYPALISHIRING", 4) << '\n'; // PINALSIGYAHRPI
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(n)
*/

File: 022_Zigzag Conversion.py
"""
Zigzag Conversion

Convert string to zigzag pattern on numRows and read row by row.
"""

def convert(s: str, numRows: int) -> str:
    if numRows <= 1 or numRows >= len(s):
        return s
    rows = [''] * min(numRows, len(s))
    cur = 0
    down = True
    for c in s:
        rows[cur] += c
        if cur == 0:
            down = True
        elif cur == numRows - 1:
            down = False
        cur += 1 if down else -1
    return ''.join(rows)


if __name__ == '__main__':
    print('Zigzag 1:', convert('PAYPALISHIRING', 3))
    print('Zigzag 2:', convert('PAYPALISHIRING', 4))

# Time Complexity: O(n)
# Space Complexity: O(n)
